tests/framework/framework.go

```diff
@@ -0,0 +1,242 @@
+	NsPrefix string
+	//  k8s client
+	K8sClient *kubernetes.Clientset
+	// cdi client
+	CdiClient *cdiClientset.Clientset
+	// generated/unique ns per test
+	Namespace *v1.Namespace
+	// generated/unique secondary ns for testing across namespaces (eg. clone tests)
+	Namespace2 *v1.Namespace // note: not instantiated in NewFramework
+	// list of ns to delete beyond the generated ns
+	namespacesToDelete []*v1.Namespace
+	// test run-time flags
+	KubectlPath  string
+	CdiInstallNs string
+	KubeConfig   string
+	Master       string
+}
+
+// TODO: look into k8s' SynchronizedBeforeSuite() and SynchronizedAfterSuite() code and their general
+//       purpose test/e2e/framework/cleanup.go function support.
+
+// initialize run-time flags
+func init() {
+fmt.Printf("\n******************* init ***********\n")
+	//flag.StringVar(&kubectlPath, "kubectl-path", "", "Set path to kubectl binary")
+	//flag.StringVar(&cdiInstallNs, "installed-namespace", "kube-system", "Set the namespace CDI is installed in")
+	//flag.StringVar(&kubeConfig, "kubeconfig", "", "absolute path to the kubeconfig file")
+	//flag.StringVar(&master, "master", "", "master url")
+	kubectlPath = flag.String("kubectl-path", "", "Set path to kubectl binary")
+	cdiInstallNs = flag.String("installed-namespace", "kube-system", "Set the namespace CDI is installed in")
+	kubeConfig = flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
+	master = flag.String("master", "", "master url")
+}
+
+// NewFramework makes a new framework and sets up the global BeforeEach/AfterEach's.
+// Test run-time flags are parsed and added to the Framework struct.
+func NewFramework(prefix string) *Framework {
+	f := &Framework{
+		NsPrefix: prefix,
+	}
+
+	// handle run-time flags
+	if !flag.Parsed() {
+		flag.Parse()
+fmt.Printf("\n********* flags:\nNFlag=%d, NArg=%d, Args=%+v\n", flag.NFlag(), flag.NArg(), flag.Args())
+flag.Visit(func(f *flag.Flag) {
+	fmt.Printf("\n**** flag=%q, val=%q\n", f.Name, f.Value.String())
+})
+fmt.Printf("\n****** lookup: kubectl-path=%q\n", flag.Lookup("kubectl-path").Value)
+		if *kubectlPath == "" {
+			Fail("flag `kubectl-path` was not specified")
+		}
+		f.KubectlPath = *kubectlPath
+		if *cdiInstallNs == "" {
+			Fail("flag `installed-namespace` was not specified")
+		}
+		f.CdiInstallNs = *cdiInstallNs
+		if *kubeConfig == "" {
+			Fail("flag `kubeconfig` was not specified")
+		}
+		f.KubeConfig = *kubeConfig
+		if *master == "" {
+			Fail("flag `master` was not specified")
+		}
+		f.Master = *master
+fmt.Printf("\n******framework:\nf.KubectlPath=%q, f.CdiInstallNs=%q, f.KubeConfig=%q, f.Master=%q",f.KubectlPath, f.CdiInstallNs, f.KubeConfig, f.Master)
+	}
+
+	BeforeEach(f.BeforeEach)
+	AfterEach(f.AfterEach)
+
+	return f
+}
+
+func (f *Framework) BeforeEach() {
+	// generate unique primary ns (ns2 not created here)
+	By(fmt.Sprintf("Building a %q namespace api object", f.NsPrefix))
+	ns, err := f.CreateNamespace(f.NsPrefix, map[string]string{
```

Comment:
Creating a NS using the K8sClient before defining the K8sClient is likely not going to work.

Reply:
Good catch, my bad!